src/Elastic.Apm/OpenTelemetry/ElasticActivityListener.cs

```diff
@@ -0,0 +1,173 @@
+		internal readonly ConcurrentDictionary<string, Transaction> ActiveTransactions = new();
+
+		internal ElasticActivityListener(IApmAgent agent, Tracer tracerInternal)
+		{
+			_logger = agent.Logger?.Scoped(nameof(ElasticActivityListener));
+			_tracer = tracerInternal;
+			listener = new ActivityListener
+			{
+				ActivityStarted = ActivityStarted,
+				ActivityStopped = ActivityStopped,
+				ShouldListenTo = _ => true,
+				Sample = (ref ActivityCreationOptions<ActivityContext> _) => ActivitySamplingResult.AllData
+			};
+			ActivitySource.AddActivityListener(listener);
+		}
+
+		private readonly IApmLogger _logger;
+		internal Tracer _tracer;
+
+
+		private Action<Activity> ActivityStarted =>
+			activity =>
+			{
+				_logger.Trace()?.Log($"ActivityStarted: name:{activity.DisplayName} id:{activity.Id} traceId:{activity.TraceId}");
+
+				if (KnownListeners.KnownListenersList.Contains(activity.DisplayName))
+					return;
+
+				Transaction transaction = null;
+				if (activity.Context != null && activity.Context.IsRemote && activity.ParentId != null)
+				{
+					var dt = TraceContext.TryExtractTracingData(activity.ParentId.ToString(), activity.Context.TraceState);
+
+					transaction = _tracer.StartTransactionInternal(activity.DisplayName, "unknown",
+						TimeUtils.ToTimestamp(activity.StartTimeUtc), true, activity.SpanId.ToString(),
+						distributedTracingData: dt);
+				}
+				else if (activity.ParentId == null)
+				{
+					transaction = _tracer.StartTransactionInternal(activity.DisplayName, "unknown",
+						TimeUtils.ToTimestamp(activity.StartTimeUtc), true, activity.SpanId.ToString(),
+						activity.TraceId.ToString());
+				}
+				else
+				{
+					Span newSpan;
+					if (_tracer.CurrentSpan == null)
+					{
+						newSpan = (_tracer.CurrentTransaction as Transaction)?.StartSpanInternal(activity.DisplayName, "unknown",
+							timestamp: TimeUtils.ToTimestamp(activity.StartTimeUtc), id: activity.SpanId.ToString());
+					}
+					else
+					{
+						newSpan = (_tracer.CurrentSpan as Span)?.StartSpanInternal(activity.DisplayName, "unknown",
+							timestamp: TimeUtils.ToTimestamp(activity.StartTimeUtc), id: activity.SpanId.ToString());
+					}
+
+					if (newSpan != null)
+					{
+						newSpan.Otel = new OTel();
+						newSpan.Otel.SpanKind = activity.Kind.ToString();
+
+						if (activity.Id != null) ActiveSpans.TryAdd(activity.Id, newSpan);
+					}
+				}
+
+				if (transaction != null)
+				{
+					transaction.Otel = new OTel();
+					transaction.Otel.SpanKind = activity.Kind.ToString();
+
+					if (activity.Kind == ActivityKind.Server)
+						transaction.Type = ApiConstants.TypeRequest;
+					else if (activity.Kind == ActivityKind.Consumer) transaction.Type = ApiConstants.TypeMessaging;
+
+					if (activity.Id != null) ActiveTransactions.TryAdd(activity.Id, transaction);
+				}
+			};
+
+
+		private Action<Activity> ActivityStopped =>
+			activity =>
+			{
+				if (activity == null)
```

Comment:
Is this ever true?

Reply:
I remember adding this during testing. Unfortunately I don't remember the details anymore, but I saw `null`s here.